Isotherms mesh refinement plugin

MFraters opened the pull request:
This plugin adds the possibility to define a minimum and maximum level of mesh refinement between two isotherms (absolute temperatures).

bangerth commented on include/aspect/mesh_refinement/isoterms.h:
```diff
@@ -0,0 +1,80 @@
+	std::vector<std::string> n_isoterms_inner;
+	unsigned int minimum_refinement_level;
+	unsigned int maximum_refinement_level;
+	std::vector<std::string> n_isoterms_outer;
```
Can you document all of these variables? You also use a non-standard indentation for some of them.

bangerth commented on source/mesh_refinement/isoterms.cc:
```diff
@@ -0,0 +1,343 @@
+                                              "isoterms",
+                                              "A mesh refinement criterion that ensures a "
+                                              "maximum and minimum refinement level between"
+                                              "two temperatures (isoterms), with the posibility"
```
The compiler will concatenate all of these strings as you write them. In other words, you probably want a space at the end (or beginning, though our convention is the end) of each of these lines so that the output has spaces between words.

bangerth commented:
I need to leave for home and so won't get to review the rest of the code, but:

1/ The correct spelling is "isotherm", not "isoterm". Can you fix the name of the file, of the class and all other occurrences of the word?

2/ Your indentation is all over the place. We have an indentation script that automatically fixes this. To this end, you need to download and install (easy) the astyle program: http://astyle.sourceforge.net/ You will need version 2.04 (not the newer version 2.05). Then run
```
  astyle options=doc/astyle.rc include/.../isotherm.h source/.../isotherm.cc
```
and commit the result. This makes sure we use a uniform indentation style everywhere.

Drop us a note here once you're done and I can resume my review.

MFraters commented:
Thanks for the quick comments! I fixed several things to do with style, formatting and variable names and their use, and added some comments on the function of the used variables.

tjhei commented:
Hey Menno,

your pull request still has many typos in comments and description strings. Before we go into more detail, I would like to ask a few questions:

1. What is exclude_composition supposed to do? I understand that your plugin sets minimum and maximum refinement levels based on the current temperature in that cell.

2. I wonder if we should choose a more flexible approach where there are FunctionParser expressions that are evaluated on each cell. As an example
```
set min = if (0<T && T<1000, 0, if(T>2000, 4, 0))
set max = if (0<T && T<1000, max, if(T>2000, 4, max)) 
```
This would allow arbitrary expressions that include computations ("min+1") and additional checks (we could export other quantities than temperature as variables!).
If the syntax is too complicated with these nested ifs (note that the syntax is ```if(condition,then-expression, else-expression)```), we can allow more than one expression to be evaluated.

MFraters commented:
Hey Timo,

The exclude composition has been put in to there because I needed the code also to work with a sticky air layer, where I don't need such a high level of refinement. So with this I can prevent the sticky air layer from refining to a very high level just because it is very cold. To make it a bit more general, I could change it to a list of numbers, instead of just a number.

I have also been thinking on generalizing this plugin to a sort of isocontour plugin, which can be used with any field within aspect by adding a parameter to each line in the input file which states on which field you want to operate. I must say that I do not know the exact workings FunctionParser well enough to know if a function parser would give additional flexibility compared to adding this functionality to the parameter list of the current plugin, but if it does it would be an interesting idea to switch to that. If this would be the approach, I would definitely be in favour of allowing multiple expressions to be evalued to prevent complicated nested if statements. 

Just as a side note, it is in the current code already possible to use both absolute and relative refinement levels by using numbers or "min", "min+1", "max-1", "max-5", etc.

tjhei commented:
> So with this I can prevent the sticky air layer from refining to a very high level just because it is very cold.

So your condition is something like ```if (c_air < 0.5 && 0 < T && T < 1000, 4, min)```?

> Just as a side note, it is in the current code already possible to use both absolute and relative refinement levels by using numbers or "min", "min+1", "max-1", "max-5", etc.

Yes, I noticed that of course. If you use the function parser you don't have to implement that logic (and stuff like ```1+min```, spaces, or other formulas "just work").

tjhei commented:
@bangerth Do you have an opinion on this? Is Menno's plugin generic enough that it is useful or should we try to generalize it?

bangerth commented on include/aspect/mesh_refinement/isotherms.h:
```diff
@@ -0,0 +1,88 @@
+         * Declare a variable which indicates how many isotherms the user has declared, and a variable to store all those isotherms
+         */
+        int number_of_isotherms;
+	std::vector<std::vector<std::string> > isotherms;
```
This is duplicative. You can always get the number of isotherms via `isotherms.size()` and this is preferable to storing the same information twice (because invariable these variables will eventually get out of synch as code is being developed further).

bangerth commented on include/aspect/mesh_refinement/isotherms.h:
```diff
@@ -0,0 +1,88 @@
+        /**
+         * Declare a variable which indicates if and which compositional field should be excluded from the process.
+         */
+        signed int exclude_composition; 
```
`int` is always signed. Do you reserve a special value for "no compositions excluded"? If so, can you document it?

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,343 @@
+        "isotherms'. The first parameter takes the "
+        "number of the compositional field and excludes "
+        "it from the min/max refinment level prodedure. "
+        "The second parameters takes a list of isoterm "
```
isoterm -> isotherm

bangerth replied:
same below -- maybe just do a global search/replace

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,343 @@
+        "available: 'Exclude composition' and 'List of "
+        "isotherms'. The first parameter takes the "
+        "number of the compositional field and excludes "
+        "it from the min/max refinment level prodedure. "
```
refinment -> refinement

I don't think I understand this sentence, though. The procedure is applied to the mesh, not to a compositional field. How can you exclude a *compositional field* from the procedure?

bangerth replied:
Also, you may want to describe what "the procedure" actually is.

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,343 @@
+        "refiment level for this isoterm, the third "
+        "subparameter sets the minimum temperature for "
+        "this isoterm and the fourth subparameter sets "
+        "the maximum termperature for the isoterm. "
```
This is a bit backward: the parameter is called `List of isotherms`, so I would suggest that the *isotherms* that define the region come first, followed by the minimal and maximal refinement levels.

As a matter of taste, I think that the parameter actually defines a `List of thermal regions`, as it (i) specifies lower and upper bounds for a region, and (ii) does so via temperatures. Or do you adjust the temperature to be an *isotherm* somewhere. In any case, you may want to consider the naming of the parameter (and plugin).

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,343 @@
+        "The minimum and maximum level can be indicated "
+        "by the absolute number of the refinement level "
+        "or by using the key words 'min' or 'max', "
+        "corrosponding respecively to the set minimum "
```
corrosponding -> corresponding

bangerth replied:
respecively -> respectively

jdannberg commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,343 @@
+        "isotherms",
+        "A mesh refinement criterion that ensures a "
+        "maximum and minimum refinement level between "
+        "two temperatures (isotherms), with the posibility "
```
posibility -> possibility or option

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,343 @@
+        "and maximum refinment level. The keywords "
+        "'min' and 'max' may also be used in combination "
+        "with addition or substraction, e.g. 'max-1' "
+        "or 'min+1'. For formating it is usefull to "
```
formating -> formatting
usefull -> useful

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,343 @@
+        "with addition or substraction, e.g. 'max-1' "
+        "or 'min+1'. For formating it is usefull to "
+        "make use of the build in option of ASPECT "
+        "to split parameters over several line by "
```
build in -> built-in
line -> lines

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,343 @@
+        "or 'min+1'. For formating it is usefull to "
+        "make use of the build in option of ASPECT "
+        "to split parameters over several line by "
+        "putting ' \\' at the end of the line. "
```
Things here will be escaped more than one: when going from source code to internal string representations, and again into latex. Just say "putting a backslash at the end of the line."

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,343 @@
+        "set List of isotherms = "
+        "max,\t max,\t 0,\t 1000; \\ \n\t\t 5,\t max-1,\t"
+        " 1000,\t 1500; \\ \n\t\t min,\t max-1,\t 1500,\t"
+        " 2000")
```
I think this piece of text would actually go better with the description of the parameter in `declare_entry`.

bangerth commented on include/aspect/mesh_refinement/isotherms.h:
```diff
@@ -0,0 +1,88 @@
+/*
+  Copyright (C) 2011, 2012 by the authors of the ASPECT code.
```
adjust the date

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,343 @@
+void
+Isotherms<dim>::tag_additional_cells () const
+{
+    if(this->get_dof_handler().n_dofs() != 0) {
```
How come you need this? When would this be called without dofs?

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,343 @@
+                        if(isotherms[it1][0].compare(0,4,"max-")==0) {
+                            std::vector<std::string> tmpNumber = Utilities::split_string_list(isotherms[it1][0],'-');
+                            isotherms_i[it1][0]=maximum_refinement_level-Utilities::string_to_int(tmpNumber[1]);
+                        } else {
```
This kind of code is inefficient: you are splitting strings and interpreting strings in the innermost loop over cells. This would be better handled (and lead to easier to read code) if you did it in `parse_parameters()` already, or if you need to refer to variables you don't know yet at that time, at the top of this function before you loop over cells.

bangerth commented:
@tjhei : I think there is relatively little harm accepting plugins. They present little maintenance burden and may serve others as examples of how to program what they want to implement themselves. I'd just like to get it into good enough shape that it can serve this role.

MFraters commented:
Thanks for the many comments (and sorry for the spelling issues)! This really helps me to write better code in the future :) 

I am doubting myself what would be a better approach. I was already planning to make a more general plugin sometime, but I wanted to share this in the mean time. Since it it now already available to people who are interested (in the form of this pull request), I am thinking of leaving it in this state until I have the time to finish a proper general version of this plugin.

bangerth commented:
It's your choice. I'm sure you can get the plugin into a shape that looks good to us with not too much work.

As far as the many comments are concerned: recall that both Timo and I started as new programmers as well at some point. We just have a few years of a head start, but believe me: we've both made all of the same mistakes at least once before as well. My observation with writing patch reviews for years now is that it gets easier very quickly: you'll understand the style we're looking for quickly, and after half a dozen review sessions of patches, you'll write patches that look very much like the ones we write -- people learn quickly :-)

bangerth commented:
@MFraters : We're not quite sure here whether you still intended to work on this, or whether you'd rather keep it local to your own ASPECT version? We can of course also discuss this next week at the hackathon.

MFraters commented:
Hey Wolfgang,
I am still planning to work on it, but I am currently testing the averaging schemes you recently implemented. I hope to have a new proposal submitted before I arrive at the hackaton, so we can discuss it.

MFraters commented:
Hey Wolfgang and Timo,
I am currently thinking on what method I should use to process the input, and I would like to hear your opinion on this. I currently see two options. The first option is in the same way which this plugin is written with a list of conditions. The second approach would be to use the function parser as Timo suggested. Here is a preview (probably overcomplicated, but showing all things I could imagine what people would actually want to do) of the two approaches which would result in the same outcome:

option 1:
```
# 						res: min & max, 	Condition 1: min & max, 			Condition 2: min &max;
Set list of fields 	= 	max-2	& max,		Temperature:	0 		& 1000; \
						max-1	& max,		Temperature:	1000 	& 1200,		Density: 2900 & 3500; \
						min+1 	& max,		Strain rate:	1e-13 	& 9e-13; \
						max 	& max, 		Density:		2900 	& 3300; \
						min 	& min, 		Pressure: 		0 		& 5e9; \
						min 	& max-1,	Viscosity:		1e23 	& 1e24; \
						min 	& max, 		C4:				0.5 	& 1;
```
option 2:
```
Set min = 	if(C4>0.5 				&& C4 < 1, 													min, \
			if(Temperature >0 		&& Temperature <0, 											max-2, \
			if(Temperature >1000 	&& Temperature <1200 	&& Density >2900 && Density <3500, 	max-1, \
			if(Density >2900 		&& Density <3300, 											max, \
			if(Pressure >0 			&& Pressure <5e9, 											min, \
			if(Viscosity >1e23 		&& Viscosity < 1e24, 										min, \
			if(Strain_rate > 1e-13 	&& Strain_rate < 9e-13, 									min+1,min)))))))

Set max = 	if(C4>0.5 				&& C4 < 1, 													max, \
			if(Temperature >0 		&& Temperature <0, 											max, \
			if(Temperature >=1000 	&& Temperature <1200 	&& Density >2900 && Density <3500, 	max, \
			if(Density >2900 		&& Density <3300, 											max, \
			if(Pressure >0 			&& Pressure <5e9, 											min, \
			if(Viscosity >1e23 		&& Viscosity < 1e24, 										max-1, \
			if(Strain_rate > 1e-13 	&& Strain_rate < 9e-13, 									max, max)))))))
```
Option 2 is in theory more flexible in terms of capabilities (although I can't imagine what more you would want to do), but it is also less readable. Option 1 also has the advantage that changing the order of the items is very straightforward, while option 2 with all the if statements might prove to be more difficult. Option 2 has the advantage that it will probably require a small amount of code, using the function parser, while option is will contain more and more complex code.

Do you have a preference for one of the two approaches?

Many thanks,

Menno

bangerth commented:
@MFraters -- what is the current status of this? You still like it? Should I take another look?

At the time when you originally proposed this patch, there was one failing test. You may want to rebase the patch. There are also a bunch of comments still un-addressed from the above.

MFraters commented:
I have been thinking about what to do with this pull request. I think that this pull request should (after some small fixes) be merged as is. I (and some of my students) actually already use it quite a lot. 

The proposal to generalize this pull request are very interesting, but I currently do not have the time to do this. And since it is just a plugin, we can always later create a isocountour refinement plugin which generalises this.

bangerth commented:
OK. Let's see if this even still works. Can you rebase and push again?

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,378 @@
+                            );
+          prm.declare_entry ("List of isotherms", "",
+                             Patterns::List(Patterns::List(Patterns::Anything(),0,100000000,","),0,100000000,";"),
+                             "Number of mininum and maximum areas defined by isotherms are going to be defined."
```
Both of these parameters could use a more extensive description. I don't think I understand what they are supposed to indicate :-)

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,378 @@
+      prm.enter_subsection("Mesh refinement");
+      {
+        minimum_refinement_level = Utilities::string_to_int(prm.get ("Minimum refinement level"));
+        maximum_refinement_level = Utilities::string_to_int(prm.get("Initial global refinement")) + Utilities::string_to_int(prm.get("Initial adaptive refinement"));
```
You can do this shorter if you do
```
  minimum_refinement_level = prm.get_int ("Minimum refinement level");
```

bangerth replied:
Still applicable.

bangerth replied:
By the way, is this really the maximum refinement level? I would have imagined that we allow refinement beyond that...

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,378 @@
+          exclude_composition = Utilities::string_to_int(prm.get("Exclude composition"));
+          std::vector<std::string> isotherms_outer_loop = Utilities::split_string_list(prm.get ("List of isotherms"),';');
+          // process the List of isotherms to get all data out of the structure
+          isotherms.resize(isotherms_outer_loop.size(),std::vector<std::string>(4,"0"));
```
All of the conversions from string to double I mentioned above should already be happening in this function.

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,378 @@
+              isotherms[isotherm_line][2]=isotherms_inner_loop[2];
+              isotherms[isotherm_line][3]=isotherms_inner_loop[3];
+
+            }
```
no need for the empty line

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,378 @@
+                                              "A mesh refinement criterion that ensures a "
+                                              "maximum and minimum refinement level between "
+                                              "two temperatures (isotherms), with the possibility "
+                                              "to exclude a composition from this criterion. "
```
I don't think I understand what "excluding a composition from this criterion" means. Do you want to say something like "with the possibility of excluding all cells from this criterion for which a particular compositional field has particular values"?

bangerth replied:
I still think that it would be quite useful to expand on what exactly the exclusion does.

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,378 @@
+                                              "available: 'Exclude composition' and 'List of "
+                                              "isotherms'. The first parameter takes the "
+                                              "number of the compositional field and excludes "
+                                              "it from this plugin. "
```
same here

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,378 @@
+                                              "parameters for one isotherm separated by a ';'. "
+                                              "Each line of isotherm parameters contains four "
+                                              "subparameters. The first subparameter sets "
+                                              "the minimun refinement level for this isotherm, "
```
-> minimum

bangerth replied:
I also don't quite understand what "minimum refinement level for this isotherm" means. An isotherm is a two-dimensional surface in a three-dimensional model. It is not a cell. Do you mean "minimum refinement level for all cells whose temperature (at at least one quadrature points) lies between the value of this and the following isotherm temperature" or similar?

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,378 @@
+                                              "'min' and 'max' may also be used in combination "
+                                              "with addition or substraction, e.g. 'max-1' "
+                                              "or 'min+1'. For formatting it is useful to "
+                                              "make use of the build-in option of ASPECT "
```
-> built-in

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,378 @@
+                                              "with addition or substraction, e.g. 'max-1' "
+                                              "or 'min+1'. For formatting it is useful to "
+                                              "make use of the build-in option of ASPECT "
+                                              "to split parameters over several line by "
```
-> lines

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,378 @@
+                                              "Example input format of List of isotherms:\n"
+                                              "set List of isotherms = "
+                                              "max,\t max,\t 0,\t 1000; \\ \n\t\t 5,\t max-1,\t"
+                                              " 1000,\t 1500; \\ \n\t\t min,\t max-1,\t 1500,\t"
```
I think you can just format this little example using the `lstlisting` environment we use in the manual.

bangerth replied:
have you checked how this is rendered in the manual?

MFraters commented on include/aspect/mesh_refinement/isotherms.h:
```diff
@@ -0,0 +1,87 @@
+         * Declare a variable to store the global minimum refinement level and a variable to store the global maximum refinement level set by the user
+         */
+        unsigned int minimum_refinement_level;
+        unsigned int maximum_refinement_level;
```
This is just to help the plugin, not the complete goal of the plugin

MFraters commented:
It is ready for another review! :)

MFraters commented:
The tester is now also happy :)

gassmoeller commented:
Hi Menno,
I will take another look, but please figure out a way to make the tests smaller. I do not want to merge 500,000 lines of test output :wink:

MFraters commented:
haha, sure, I will bring it down a bit :)

MFraters commented:
It is ready for review.

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,373 @@
+                             Patterns::Double (),
+                             "This option allows to exclude one composition from being condidered by this plugin. "
+                             "This can for example be used to exclude the composition representing a sticky air, which "
+                             "is as cold as the surface, but you do not want to be refined."
```
You should say somewhere that this variable is the *index* of one compositional field that allows to disable the plugin, *if the compositional field has a value less than 0.5*.

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,373 @@
+      {
+        prm.enter_subsection("Isotherms");
+        {
+          prm.declare_entry ("Exclude composition", "-1",
```
maybe make this the string equivalent of `numbers::invalid_unsigned_int`?

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,373 @@
+                            );
+          prm.declare_entry ("List of isotherms", "",
+                             Patterns::List(Patterns::List(Patterns::Anything(),0,100000000,","),0,100000000,";"),
+                             "A list of isotherms separated by semi-colins (;). Each isotherm entry consists of "
```
```suggestion
                             "A list of isotherms separated by semi-colons (;). Each isotherm entry consists of "
```

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,373 @@
+                             Patterns::List(Patterns::List(Patterns::Anything(),0,100000000,","),0,100000000,";"),
+                             "A list of isotherms separated by semi-colins (;). Each isotherm entry consists of "
+                             "four values separted by a comma. The first two values indicate the minimum and maximum "
+                             "refinement levels respecitvely. The can be an integer, but also the words min or max "
```
```suggestion
                             "refinement levels respecitvely. These can be an integer, but also the words min or max "
```

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,373 @@
+    void
+    Isotherms<dim>::parse_parameters (ParameterHandler &prm)
+    {
+      std::vector<std::pair<std::string,std::string> > isotherms_raw;
```
move the declaration down to where you first use the variable

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,373 @@
+        prm.enter_subsection("Isotherms");
+        {
+          exclude_composition = Utilities::string_to_int(prm.get("Exclude composition"));
+          std::vector<std::string> isotherms_outer_loop = Utilities::split_string_list(prm.get ("List of isotherms"),';');
```
can you make the variable `const`?

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,373 @@
+          // loop through all given isotherms
+          for ( unsigned int isotherm_line = 0; isotherm_line != isotherms_raw.size(); ++isotherm_line )
+            {
+              std::vector<std::string> isotherms_inner_loop = Utilities::split_string_list(isotherms_outer_loop[isotherm_line]);
```
can you make the variable `const`?

also, split to shorter lines

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,373 @@
+              isotherms_raw[isotherm_line].second = isotherms_inner_loop[1];
+              // The temperatures for the min and max isotherms respectively
+              isotherms[isotherm_line].first  = Utilities::string_to_double(isotherms_inner_loop[2]);
+              isotherms[isotherm_line].second = Utilities::string_to_double(isotherms_inner_loop[3]);
```
what happens if you enter something in the input file that is not actually a number?

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,373 @@
+        {
+          if (isotherms_raw[isotherm_line].first.compare(0,3,"max") == 0)
+            {
+              if (isotherms_raw[isotherm_line].first.compare(0,4,"max-") == 0)
```
here and below, you are assuming that a user actually writes `max-2`, but not `max - 2` or similar. That's ok, but I think you need to do some error checking here. For example, the if you entered `max - 2`, you'd fall into the `else` branch below where you just assume that the user had written `max`.

bangerth commented on source/mesh_refinement/isotherms.cc:
```diff
@@ -0,0 +1,373 @@
+                                              " 1000,\t 1500; \\ \n\t\t min,\t max-1,\t 1500,\t"
+                                              " 2000"
+                                              "\\end{lstlisting}"
+                                              "The criterion is checked for every quaderature point "
```
this seems excessive. don't you think it would be enough to do this only at the midpoint of the cell?

anne-glerum commented:
Hi @MFraters, we think this mesh refinement criterion could prove very useful to us. Will you come back to it in the foreseeable future?

MFraters commented:
Hey Anne,
Sorry for the late reply, I had to think a bit about that. I do not think I will be able to finish this before the hackaton, but I am planning to spend time on this again during the hackaton and hope to (finally) finish it there. Is that early enough for you?

anne-glerum commented:
Ok, thanks, that would be great Menno! It would be really useful to focus resolution in our rift models.

On 17 Jun 2020, at 18:23, Menno Fraters <notifications@github.com<mailto:notifications@github.com>> wrote:



Hey Anne,
Sorry for the late reply, I had to think a bit about that. I do not think I will be able to finish this before the hackaton, but I am planning to spend time on this again during the hackaton and hope to (finally) finish it there. Is that early enough for you?

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://github.com/geodynamics/aspect/pull/295#issuecomment-645455548>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AB2HGCAX3O2PKLVT6RZVU43RXDUWNANCNFSM4A7YPKQA>.

MFraters commented:
closed by #3725 :tada:
